src/lib.rs

```diff
@@ -110,36 +106,27 @@ pub struct Queue<T> {
     data: T,
 }
 
-
-impl <T: Send> Queue<T> {
-    /// Creates a new, uninitialized, `Queue`.
-    pub fn new(data: T) -> Queue<T> {
-        return Queue {
-            qh : std::ptr::null_mut(),
-            qqh : std::ptr::null_mut(),
-            cb: None,
-            data: data,
-        };
+impl<T> Drop for Queue<T> {
+    fn drop(&mut self) {
+        unsafe { nfq_close(self.qh) };
     }
+}
 
-    /// Opens a NFLOG handler
+impl <T: Send> Queue<T> {
+    /// Creates a new Queue and opens a NFLOG handler
     ///
     /// This function obtains a netfilter queue connection handle. When you are
     /// finished with the handle returned by this function, you should destroy it
     /// by calling `close()`.
     /// A new netlink connection is obtained internally
     /// and associated with the queue connection handle returned.
-    pub fn open(&mut self) {
-        self.qh = unsafe { nfq_open() };
-    }
-
-    /// Closes a NFLOG handler
-    ///
-    /// This function closes the nfqueue handler and free associated resources.
-    pub fn close(&mut self) {
-        assert!(!self.qh.is_null());
-        unsafe { nfq_close(self.qh) };
-        self.qh = std::ptr::null_mut();
+    pub fn new(data: T) -> Queue<T> {
+        Queue {
+            qh: unsafe { nfq_open() },
+            qqh: std::ptr::null_mut(),
+            cb: None,
+            data,
+        }
```

Comment:
This function should test the return value of `nfq_open`, and return a `Result` to handle errors. If `nfq_open` fails here, other calls will cause undefined behavior (not visible since the `assert` calls are removed)

Reply:
I agree, that is what I would have liked to do. I didn't want to introduce an error handling pattern without your input though. Are you happy with a manual implementation of an Error enum?

Reply:
Sure, that is usually what I use. For ex, see https://github.com/rusticata/pcap-parser/blob/master/src/error.rs